sql/src/main/java/org/apache/druid/sql/calcite/schema/MetadataSegmentView.java

```diff
@@ -0,0 +1,255 @@
+  public void start()
+  {
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+    try {
+      if (isCacheEnabled) {
+        scheduledExec.schedule(new PollTask(), pollPeriodInMS, TimeUnit.MILLISECONDS);
+      }
+      lifecycleLock.started();
+      log.info("MetadataSegmentView Started.");
+    }
+    finally {
+      lifecycleLock.exitStart();
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    if (!lifecycleLock.canStop()) {
+      throw new ISE("can't stop.");
+    }
+    log.info("MetadataSegmentView is stopping.");
+    if (isCacheEnabled) {
+      scheduledExec.shutdown();
+    }
+    log.info("MetadataSegmentView Stopped.");
+  }
+
+  private void poll()
+  {
+    log.info("polling published segments from coordinator");
+    final JsonParserIterator<DataSegment> metadataSegments = getMetadataSegments(
+        coordinatorDruidLeaderClient,
+        jsonMapper,
+        responseHandler,
+        segmentWatcherConfig.getWatchedDataSources()
+    );
+
+    final DateTime timestamp = DateTimes.nowUtc();
+    while (metadataSegments.hasNext()) {
+      final DataSegment interned = DataSegmentInterner.intern(metadataSegments.next());
+      // timestamp is used to filter deleted segments
+      publishedSegments.put(interned, timestamp);
+    }
+    // filter the segments from cache whose timestamp is not equal to latest timestamp stored,
+    // since the presence of a segment with an earlier timestamp indicates that
+    // "that" segment is not returned by coordinator in latest poll, so it's
+    // likely deleted and therefore we remove it from publishedSegments
+    // Since segments are not atomically replaced because it can cause high
+    // memory footprint due to large number of published segments, so
+    // we are incrementally removing deleted segments from the map
+    // This means publishedSegments will be eventually consistent with
+    // the segments in coordinator
+    publishedSegments.entrySet().removeIf(e -> e.getValue() != timestamp);
+    cachePopulated.set(true);
+  }
+
+  public Iterator<DataSegment> getPublishedSegments()
+  {
+    if (isCacheEnabled) {
+      Preconditions.checkState(
+          lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS) && cachePopulated.get(),
+          "hold on, still syncing published segments"
+      );
+      return publishedSegments.keySet().iterator();
+    } else {
+      return getMetadataSegments(
+          coordinatorDruidLeaderClient,
+          jsonMapper,
+          responseHandler,
+          segmentWatcherConfig.getWatchedDataSources()
+      );
+    }
+  }
+
+  // Note that coordinator must be up to get segments
+  private JsonParserIterator<DataSegment> getMetadataSegments(
+      DruidLeaderClient coordinatorClient,
+      ObjectMapper jsonMapper,
+      BytesAccumulatingResponseHandler responseHandler,
+      Set<String> watchedDataSources
+  )
+  {
+    String query = "/druid/coordinator/v1/metadata/segments";
```

Comment:
Would be nice to do something like this: #7390

Reply:
ok, will check that, thanks.